src/sync/app_data.py

```diff
@@ -55,73 +55,44 @@ def num_records(self) -> int:
         )
         return len(self._found)
 
-    def _handle_new_records(self, max_retries: int) -> None:
+    async def _handle_new_records(self, max_retries: int) -> None:
         # Drain the dune_results into "found" and "not found" categories
-        while self.new_rows:
-            row = self.new_rows.pop()
-            app_hash = row["app_hash"]
-            cid = Cid(app_hash)
-            app_data = cid.get_content(max_retries)
-
-            # Here it would be nice if python we more like rust!
-            if app_data is not None:
-                # Row is modified and added found items
-                log.debug(f"Found content for {app_hash} at CID {cid}")
-                row["content"] = app_data
-                self._found.append(row)
-            else:
-                # Unmodified row added to not_found items
-                log.debug(
-                    f"No content found for {app_hash} at CID {cid} after {max_retries} retries"
-                )
-                # Dune Records are string dicts.... :(
-                row["attempts"] = str(max_retries)
-                self._not_found.append(row)
+        self._found, self._not_found = await Cid.fetch_many(self.new_rows, max_retries)
 
-    def _handle_missing_records(self, max_retries: int, give_up_threshold: int) -> None:
-        while self.missing_values:
-            row = self.missing_values.pop()
-            app_hash = row["app_hash"]
-            cid = Cid(app_hash)
-            app_data = cid.get_content(max_retries)
-            attempts = int(row["attempts"]) + max_retries
-
-            if app_data is not None:
-                log.debug(
-                    f"Found previously missing content hash {row['app_hash']} at CID {cid}"
-                )
-                self._found.append(
-                    {
-                        "app_hash": app_hash,
-                        "first_seen_block": row["first_seen_block"],
-                        "content": app_data,
-                    }
-                )
-            elif app_data is None and attempts > give_up_threshold:
+    async def _handle_missing_records(
+        self, max_retries: int, give_up_threshold: int
```

Comment:
Is the idea here for example to have a `give_up_threshold` of 9 and `max_retries` of 3 in order to try fetching a file at most 3 times with 3 retries per iteration? That would seem needlessly complex.

Instead you could write `fetch_many()` to only do one request per AppData per call (and update `attempts` accordingly). And then `_handle_missing_records()` could simply drop all AppDatas from the `missing` list that exceeded the `max_retries`. That way you should have minimum latency while only needing one `retry` parameter.

Reply:
Yes! I totally agree that this very strange approach (as you have accurately assumed it works) and I have a hunch that it does not need to be so complex. This logic is currently mimicking how we fetch app data in the [legacy setup](https://github.com/cowprotocol/dune-bridge/blob/c07bc1ff173c58e63d6ad8f89a936601c76a1eec/src/referral_maintenance.rs)

I believe the idea is the following: It is well known that the first attempt with such a low timeout=1s does need to be retried more than once at a time. However, sometimes the issue can be with the IPFS node provider being unavailable, so we don't entirely give up on a single run of the cron job, but try again on the next run loop (to a maximum). 

I would however, need to test the theory (could probably do it here) and collect some metrics on how many retries it actually takes to get the results -- or if on average recovered app content is actually recovered within the first attempt (I do not belive this to be true, because I have experienced the scenario in this program where some app data is not recovered in a single run - with 3 retrys, but is found on the next run). 

I understand that this is a bit ugly/confusing but I do think we should keep it like this. I could however collect some data to see how many retries are actually needed on average to get the app data on the first run. What do you think?

Reply:
Good to know. Would be interesting to see how often it happens that an `AppData` can be fetched at the 2nd or 3rd try within the same chronjob instance. If this number is very low one could consider to always do a single try per chronjob only to make the logic a bit less complex.

Reply:
I will do some naive benchmarking here in this unit test (by logging a warning every time attempts > 1) for a large set. Lets see how it plays out.

Reply:
On a single run with 84 records we got: `attempts => count`

```
{
	1: 72,
	2: 10,
	4: 1,
        5: 1
}
```

Reply:
Another run: `{1: 77, 2: 5, 4: 1, 5: 1}`
Another run: `{2: 82, 4: 2}`

I think the 3 retries per run seems to be pretty good at capturing this, but please let me know what you think.

Reply:
Yeah, seems alright. Probably better to not introduce unexpected problems by changing the code unnecessarily.

Reply:
So in that case I believe we can call this resolved.